File: LeetCode/p39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(const vector<int> &candidates, int target) {
        len = candidates.size();
        raw_data = candidates;
        m_target = target;
        tmp_vec.reserve(len);
        backtracking(0, 0);
        return result;
    }

private:
    int len;
    int m_target;
    vector<int> tmp_vec;
    vector<int> raw_data;
    vector<vector<int>> result;

    void backtracking(int sum, int start) {
        if (sum == m_target) {
            result.emplace_back(tmp_vec);
            return;
        }
        if (sum > m_target)
            return;
        for (int i = start; i < len; ++i) {
            tmp_vec.push_back(raw_data[i]);
            backtracking(sum + raw_data[i], i);
            tmp_vec.pop_back();
        }
    }
};

File: LeetCode/p39.rs
impl Combine {
  pub fn combine(mut self) -> Vec<Vec<i32>> {
    self.backtrack(0, 0);
    self.ret
  }

  fn backtrack(&mut self, mut start: i32, sum: i32) {
    if sum >= self.target {
      if sum == self.target {
        self.ret.push(self.tmp.clone());
      }
      return;
    }
    for i in (start as usize)..self.candidates.len() {
      self.tmp.push(self.candidates[i]);
      self.backtrack(start, sum + self.candidates[i]);
      start += 1;
      self.tmp.pop();
    }
  }
}

impl Solution {
  pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    Combine{candidates, target, tmp: Vec::new(), ret: Vec::new()}.combine()
  }
}

